compiler/rustc_trait_selection/src/solve/trait_goals.rs

```diff
@@ -505,28 +505,32 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
             return vec![];
         }
 
-        let mut unsize_dyn_to_principal = |principal: Option<ty::PolyExistentialTraitRef<'tcx>>| {
+        let mut unsize_dyn_to_principal = |super_trait_ref: Option<ty::PolyTraitRef<'tcx>>| {
             ecx.probe_candidate("upcast dyn to principle").enter(|ecx| -> Result<_, NoSolution> {
-                // Require that all of the trait predicates from A match B, except for
-                // the auto traits. We do this by constructing a new A type with B's
-                // auto traits, and equating these types.
-                let new_a_data = principal
-                    .into_iter()
-                    .map(|trait_ref| trait_ref.map_bound(ty::ExistentialPredicate::Trait))
-                    .chain(a_data.iter().filter(|a| {
-                        matches!(a.skip_binder(), ty::ExistentialPredicate::Projection(_))
-                    }))
-                    .chain(
-                        b_data
-                            .auto_traits()
-                            .map(ty::ExistentialPredicate::AutoTrait)
-                            .map(ty::Binder::dummy),
-                    );
-                let new_a_data = tcx.mk_poly_existential_predicates_from_iter(new_a_data);
-                let new_a_ty = Ty::new_dynamic(tcx, new_a_data, b_region, ty::Dyn);
-
-                // We also require that A's lifetime outlives B's lifetime.
-                ecx.eq(goal.param_env, new_a_ty, b_ty)?;
+                // Relate the principals, to make sure that the supertrait we found in
+                // the target actually satisfies the principal of the target.
+                // FIXME: This should be very easy to support "upcasting" from
+                // `dyn Trait + Send` to `dyn Send`.
+                if let Some(super_trait_ref) = super_trait_ref {
+                    ecx.eq(
+                        goal.param_env,
+                        super_trait_ref,
+                        b_data.principal().unwrap().with_self_ty(tcx, a_ty),
+                    )?;
+                }
+                // Secondly, make sure that all of the non-principal bounds of the
+                // target (associated types and auto traits) are satisfied by the source.
+                ecx.add_goals(
+                    b_data
+                        .projection_bounds()
+                        .map(|bound| goal.with(tcx, bound.with_self_ty(tcx, a_ty))),
```

Comment:
Since we're proving projection predicates here instead of just relating things structurally, that triggers the assertion down in `ReplaceProjectionWith` in `predicates_for_object_candidate` for the `tests/ui/type-alias-impl-trait/normalize-hidden-types.rs` test.

Reply:
Actually, I can reproduce this ICE on nightly:

```
// compile-flags: -Ztrait-solver=next

trait Trait {
    type Gat<'lt>;
}

fn test<T, F: FnOnce(<T as Trait>::Gat<'_>) -> S, S>() {}

fn main() {
    test::<u8, dyn FnOnce(<u8 as Trait>::Gat<'_>), _>();
}
```

The reason that this PR triggers the change is that we're now proving more predicates. I'll explain more in sync.

Reply:
please add as a test with comment